### DISCUSSION TITLE
Inconsistent behavior of DeepExplainer with multi-output models ??

### DISCUSSION THREAD

**Discussion by papsissoko** [shap/shap#4137] [Q&A]
Hello, 
According to the SHAP documentation, DeepExplainer is supposed to work only with models that produce a single tensor. This is reinforced by the assertion in the TFDeep class: "assert not isinstance(self.model_output, list), 'The output of the model to explain must be a single tensor!'". However, in practice, we have observed that Keras models with multiple outputs (i.e. multi-output regression) do not trigger this error and are processed without issues. This raises the question: - Has support for multi-output Keras models been  implemented in DeepExplainer? 
Thank you in advance for any clarification.

**Comment by DanteTrb** [comment]
Excellent question and you're absolutely right to point out this apparent inconsistency.

Let me clarify and confirm, both empirically and conceptually, what's happening with `DeepExplainer` and Keras multi-output models.

---

###  1. What the SHAP Docs Say

The SHAP documentation for `DeepExplainer` clearly states that it only supports **single-output models**, enforcing the constraint:

> "The output of the model to explain must be a single tensor."

This is internally validated by:
python
assert not isinstance(self.model_output, list)

###  2. But in Practice... It Works?
Yes, and I’ve replicated it.

Here’s a minimal working example:

```
import shap
import numpy as np
from tensorflow.keras.models import Model
from tensorflow.keras.layers import Input, Dense

# Define a simple multi-output model
inputs = Input(shape=(4,))
out1 = Dense(1)(inputs)
out2 = Dense(1)(inputs)
model = Model(inputs=inputs, outputs=[out1, out2])
model.compile(optimizer='adam', loss='mse')

#Dummy data
X = np.random.randn(20, 4)
model.fit(X, [np.random.randn(20, 1), np.random.randn(20, 1)], epochs=1)

# Attempt to explain with DeepExplainer
explainer = shap.DeepExplainer(model, X)  # No error raised!
```

This does not raise the expected assertion error, despite the output being a list of tensors.

### 3. Why Does It Work?
Likely explanation:

Internally, model(inputs) returns a list of tensors, but Keras and TensorFlow often convert or wrap outputs in ways that avoid triggering the assert under certain conditions (e.g., eager execution).

SHAP might be only using the first output, or implicitly handling the list — but without ensuring correct attribution for each head.

This can result in misleading or incomplete SHAP values, especially in multi-task learning or multi-output regression/classification.

###  4. Recommended Best Practice
To ensure safe and interpretable SHAP usage, you should explicitly split the model into per-output submodels:

```
# For output 1
model_head1 = Model(inputs=model.input, outputs=model.output[0])
explainer1 = shap.DeepExplainer(model_head1, background_data)
shap_values1 = explainer1.shap_values(test_data)

# Repeat for output 2, etc.

```
This guarantees that each explanation refers only to the corresponding output head, avoiding leakage or attribution confusion.

###  5. Suggestion to SHAP Maintainers
It may be helpful to:

Clarify in the docs whether multi-output support is explicitly supported or just coincidentally allowed,

Add a warning message when len(model.outputs) > 1, and

Indicate that only the first output may be explained if not explicitly handled.

**Comment by papsissoko** [comment]
I followed your recommendation by isolating the model to keep only a single output (e.g., t+1h), and I compared it to the original multi-output model, focusing specifically on the SHAP values generated for that same output.

To do this, I took a sample of 247 observations and extracted the SHAP values associated with the t+1h output in both cases. The result: the values are strictly identical for each sample.

There is therefore no difference in the explanations between the full model and the reduced single-output version.

Here is the code I tested:

result = []

output_first = Lambda(lambda x: x[:, 0:1], name="scalar_output")(keras_model.outputs[0])
model_head1 = Model(inputs=keras_model.inputs, outputs=output_first)

explainer1 = shap.DeepExplainer(model_head1, X_array)
shap_values1 = explainer1.shap_values(X_array)
result.append(shap_values1.squeeze(2))

explainer = shap.DeepExplainer(keras_model, X_array)
shap_values = explainer.shap_values(X_array)  # shape = (n_samples, n_features=327, predict_to)
result.append(shap_values[:, :, 0])

result = np.vstack([result[0], result[1]])
df = pd.DataFrame(result)
df.to_csv("lol.csv")

var = np.var([result[:247], result[247:]], axis=0)
moy_var = np.mean(var)
diff = result[:247] - result[247:]
diff_moy = np.mean(np.abs(diff))

X_1 = result[:247]
X_2 = result[247:]
index = range(247)

plt.plot(index, X_1, color="blue", label="isolated output 0")
plt.plot(index, X_2, color="orange", label="non-isolated output 0")
plt.xlabel("sample")
plt.ylabel("SHAP value")
plt.title(f"For 247 samples: mean_diff={diff_moy} and mean_var={moy_var}")
plt.savefig("comparaison_sorties.png", dpi=300, bbox_inches="tight")


When I plot the two curves on the same graph, they completely overlap. Also, both the mean and variance between the two sets of SHAP values are exactly zero.
